rosbag2/include/rosbag2/compression_interfaces/base_decompressor_interface.hpp

```diff
@@ -0,0 +1,75 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2__COMPRESSION_INTERFACES__BASE_DECOMPRESSOR_INTERFACE_HPP_
+#define ROSBAG2__COMPRESSION_INTERFACES__BASE_DECOMPRESSOR_INTERFACE_HPP_
+
+#include <memory>
+#include <string>
+
+#include "rosbag2/visibility_control.hpp"
+
+/**
+ * An interface for developers adding a new decompression algorithm to rosbag2. These functions
+ * must be implemented to that a reader can properly decompress a file or bag message.
+ * A corresponding compressor must also be implemented.
+ *
+ * Example file decompression usage:
+ *
+ * MyDecompressor my_decompressor();
+ * std::string current_uri = get_current_file();
+ * std::string compressed_path_uri = my_decompressor.decompress_uri(current_uri);
+ * storage = storage_factory.open_read_only(compressed_path_uri, storage_options.storage_id);
+ *
+ * Example message decompression usage:
+ *
+ * MyDecompressor my_decompressor();
+ * std::shared_ptr<SerializedBagMessage> bag_message = storage.read_next();
+ * std::shared_ptr<SerializedBagMessage> decompressed_message =
+ *   my_decompressor.decompress_serialized_bag_message(bag_message);
+ */
+namespace rosbag2
+{
+
+class ROSBAG2_PUBLIC BaseDecompressorInterface
+{
+public:
+  virtual ~BaseDecompressorInterface() {}
+
+  /**
+   * Decompress a file on disk.
+   *
+   * \param uri Input file to decompress with file extension.
+   * \return The relative path to the decompressed file without the compressed extension.
+   */
+  virtual std::string decompress_uri(const std::string & uri) = 0;
+
+  /**
+   * Decompress serialized_data in a serialized bag message.
+   *
+   * \param bag_message A serialized bag message.
+   * \return A shared pointer to the bag message with decompressed serialized_data.
+   */
+  virtual std::shared_ptr<SerializedBagMessage> decompress_serialized_bag_message(
+    std::shared_ptr<SerializedBagMessage> bag_message) = 0;
```

Comment:
same comment about constness.

Reply:
I couldn't reply to your other comment, but basically we need access to the `bag_message->serialized_data` so we can actually compress it and then replace it in the serialized bag message.

Reply:
I don't understand this. Is the intention of this function to (de)compress data in place? If so, then I'd expect this function to have a `void` or `bool` return type. Also, this should then be documented that the input parameter will be changed. Also, if being changed, I don't think a `shared_ptr` is the right pointer type (raw pointer could be more applicable if so). 
If at one point the function makes a copy of the `bag_message` I think the input parameter should be const.

Reply:
The goal of this function is to create a new `SerializedBagMessage` and deallocate the old one since we no longer need it.
We could do that in the function or outside. I prefer within the function since that way we don't put the responsibility on someone else.

On the other hand, we can keep it `const`, however, I don't see where the message would be used after it's compressed. 

Regarding using `SerializedBagMessage` specifically - We want to make it clear that compression is specifically for messages in `rosbag2`.  I don't think the goal of this compressor is to compress things outside of `rosbag2`.